### PR TITLE
Coalesce contigs into single array/partition

### PR DESCRIPTION
**Pull request by mlathara**
This adds support for GenomicsDBImporter to have multiple contigs in the same GenomicsDB partition/array. This will hopefully help import times in cases where users have many thousands of contigs.

Right now, I've made it so that contigs don't get split up between arrays. This means the arrays could be horribly skewed depending on contig size. Shouldn't be a huge deal to amend support for the more general case if we need to...

I also added a new json file to the genomicsdb_meta_dir with just the column offsets for the array. I thought about adding it to the existing row index bounds file, but it seems much simpler/easier to do it to a new file.

Lastly, based on what I've heard in the GATK meetings, it seems like it is a fairly common use case for folks to parallelize import by importing each interval to a entirely different genomicsdb workspace (presumably on a different node). This isn't really supported directly in the current implementation since we go through the vid and create the partitions under the covers instead of the user specifying the partitions. I don't know if its possible for users to mess with the vcfs/headers to only specify the contigs they want in a given interval and then import to a single array that way....Alternatively, we could make sure to introduce the htslib route of reading vcfs and make it more feasible to actually run multiple threads on a given GenomicsDBImporter process. Not really a question there I guess, but figured I'd throw it out there for comment.

### CODE REVIEW

**Comment by mlathara** [comment]
I clearly didn't spend enough time thinking about that last paragraph (parallelizing the multi-contig import). It seems easy enough to use the interval list passed in by the user to filter the contigs and only import/coalesce the contigs specified as part of that list.

I've made the code change for that, and will push soon as I've augmented the tests for that as well.

**Comment by mlathara** [comment]
OK - added the change to allow coalescing of specified contigs instead of doing all contigs.

File: src/main/cpp/include/vcf/genomicsdb_bcf_generator.h [hunk 1]
```diff
@@ -34,7 +34,7 @@
 class GenomicsDBBCFGenerator {
  public:
   GenomicsDBBCFGenerator(const std::string& loader_config_file, const genomicsdb_pb::ExportConfiguration *query_config_pb,
-                         const char* chr, const int start, const int end,
+                         const char* chr, const long start, const long end,
```
Comment (nalinigans) [inline src/main/cpp/include/vcf/genomicsdb_bcf_generator.h:37]: Did we miss this when the 64-bit support was implemented?

File: src/main/cpp/src/vcf/genomicsdb_bcf_generator.cc [hunk 2]
```diff
@@ -64,6 +64,9 @@ GenomicsDBBCFGenerator::GenomicsDBBCFGenerator(const std::string& loader_config_
     int64_t column_end = contig_info.m_tiledb_column_offset + static_cast<int64_t>(end) - 1; //since VCF positions are 1 based
     m_query_config.set_column_interval_to_query(column_begin, column_end);
   }
+  else if (start != 0 || end != 0) {
+    m_query_config.set_column_interval_to_query(start, end);
+  }
```
Comment (nalinigans) [inline src/main/cpp/src/vcf/genomicsdb_bcf_generator.cc:69]: Are `start` and `end` TileDB column offsets here or contig positions?

**Comment by nalinigans** [comment]
> OK - added the change to allow coalescing of specified contigs instead of always doing all contigs.
> 
> Another point I didn't bring out earlier...this change names the arrays/partitions based on it's rank. This will make it so previous versions of GATK won't be able to read these workspaces. Something to call out to GATK.
> 
> For now, I've chosen to use the old naming convention when not coalescing multiple contigs (so the incompatibility is only for cases that avail of the new support we're adding). Maybe that makes it easier? Though I was also pondering if we should just switch to the new naming schema for all cases....



> OK - added the change to allow coalescing of specified contigs instead of always doing all contigs.
> 
> Another point I didn't bring out earlier...this change names the arrays/partitions based on it's rank. This will make it so previous versions of GATK won't be able to read these workspaces. Something to call out to GATK.
> 
> For now, I've chosen to use the old naming convention when not coalescing multiple contigs (so the incompatibility is only for cases that avail of the new support we're adding). Maybe that makes it easier? Though I was also pondering if we should just switch to the new naming schema for all cases....

The gatk folks seem to want forward compatibility as well...

File: src/main/cpp/src/genomicsdb/variant_storage_manager.cc [hunk 3]
```diff
@@ -750,3 +773,29 @@ void VariantStorageManager::update_row_bounds_in_array(const int ad, const int64
...
+    rapidjson::Document json_doc;
+    json_doc.SetObject();
+    json_doc.AddMember("min_column", min_column, json_doc.GetAllocator());
+    json_doc.AddMember("max_column", max_column, json_doc.GetAllocator());
+    rapidjson::StringBuffer buffer;
+    rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
+    json_doc.Accept(writer);
+  
+    if (write_file(m_tiledb_ctx, column_filepath, buffer.GetString(), strlen(buffer.GetString()))) {
+      throw VariantStorageManagerException(std::string("Could not write to column bounds file ")+column_filepath
```
[the remark this answers is not in the document]
Comment (mlathara) [reply src/main/cpp/src/genomicsdb/variant_storage_manager.cc:792]: I'll change this to use logger

File: src/main/cpp/include/vcf/genomicsdb_bcf_generator.h (hunk 1 above)
Comment (mlathara) [reply src/main/cpp/include/vcf/genomicsdb_bcf_generator.h:37]: Ah - hmm, possibly? Though I never remember what level of support htsjdk has for this...I made this change to sorta overload start/end (which used to only be contig start/end) to also play as tiledb column start/end (when chr is null or blank).

File: src/main/cpp/src/vcf/genomicsdb_bcf_generator.cc (hunk 2 above)
Comment (mlathara) [reply src/main/cpp/src/vcf/genomicsdb_bcf_generator.cc:69]: see above

**Comment by mlathara** [comment]
There is no default `coalesce_to_num_partitions` per se. It is not part of the json/configuration right now. I keep the usual flow to generate `GenomicsDBImporter` and then expose a new method called `coalesceContigsIntoNumPartitions` which will get called from GATK. So this path only gets triggered when the user specifies it through a new GATK argument - otherwise we stay with the old path.

Regarding forward compatibility - from the meeting yesterday, I thought they were fine with losing that - but they did want a clear warning/error that tells the user what the issue is.

**Comment by nalinigans** [comment]
> Regarding forward compatibility - from the meeting yesterday, I thought they were fine with losing that - but they did want a clear warning/error that tells the user what the issue is.

Would it be possible to get the older gatk versions without this support throw out a clear warning? Perhaps, protobuf will take care of this...

**Comment by mlathara** [comment]
> Would it be possible to get the older gatk versions without this support throw out a clear warning? Perhaps, protobuf will take care of this...

The coalescing stuff isn't part of the config, so no protobuf involved in this case.

We'll have to run with the older versions to be see what happens exactly - but there is no danger of any sort of silent corruption and the query succeeding. It should just fail saying something like the array folder schema is incorrect. So, we'll have to add to the documentation saying that this error means you're querying a GenomicsDB workspace that was generated with a newer GATK version, and that isn't supported (not supported for some types of workspaces anyway...).

It would be nice to actually throw an error saying something like use the latest GATK to query - but unfortunately we didn't have that clairvoyance :grin:

**Comment by nalinigans** [comment]
> It would be nice to actually throw an error saying something like use the latest GATK to query - but unfortunately we didn't have that clairvoyance 😁

Agreed with the clairvoyance part... But, going forward, maybe we should build some checking in, let the command succeed if possible but clearly put out messages saying that some meta information is not understood by this version.

File: src/main/cpp/src/genomicsdb/variant_storage_manager.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mlathara) [reply src/main/cpp/src/genomicsdb/variant_storage_manager.cc]: done

**Comment by mlathara** [comment]
Swapped `throw new...` for `logger.fatal...`. Definitely will need another pass later on, but at least it's not worse than before.

**Pull request merged**